misk-cron/src/main/kotlin/misk/cron/CronManager.kt

```diff
@@ -0,0 +1,136 @@
+import java.util.PriorityQueue
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class CronManager @Inject constructor(moshi: Moshi) {
+  @Inject private lateinit var clock: Clock
+  @Inject private lateinit var jobQueue: JobQueue
+  @Inject @ForMiskCron private lateinit var queueName: QueueName
+  @Inject @ForMiskCron private lateinit var zoneId: ZoneId
+
+  private val cronJobAdapter = moshi.adapter<CronJob>()
+
+  data class CronEntry(
+    val name: String,
+    val cronTab: String,
+    val executionTime: ExecutionTime,
+    val runnable: Runnable
+  )
+
+  private val cronEntries = mutableMapOf<String, CronEntry>()
+
+  data class CronQueueEntry(val executeAt: Instant, val cronEntry: CronEntry)
+
+  private val cronQueueComparator: Comparator<CronQueueEntry> =
+    compareBy { it.executeAt.toEpochMilli() }
+  private val cronQueue = PriorityQueue(cronQueueComparator)
+
+  private var lastCronRunAt: Instant? = null
+
+  fun addCron(name: String, crontab: String, cron: Runnable) {
+    require(name.isNotEmpty()) { "Expecting a valid cron name" }
+    require(cronEntries[name] == null) { "Cron $name is already registered" }
+    logger.info { "Adding cron entry $name, crontab=$crontab" }
+
+    val cronDefinition = CronDefinitionBuilder.instanceDefinitionFor(CronType.UNIX)
+    val executionTime = ExecutionTime.forCron(CronParser(cronDefinition).parse(crontab))
+    val cronEntry = CronEntry(name, crontab, executionTime, cron)
+
+    cronEntries[name] = cronEntry
+  }
+
+  internal fun removeAllCrons() {
+    logger.info { "Removing all cron entries" }
+    cronEntries.clear()
+    cronQueue.clear()
+    lastCronRunAt = null
+  }
+
+  private fun rebuildQueue() {
+    cronQueue.clear()
+    cronEntries.values.forEach { queueNextExecution(it) }
+  }
+
+  private fun queueNextExecution(cronEntry: CronEntry) {
+    val now = ZonedDateTime.ofInstant(clock.instant(), zoneId)
+    val nextExecutionTime = cronEntry.executionTime.nextExecution(now).orElseThrow()
+      .withSecond(0)
+      .withNano(0)
+
+    logger.info { "Next execution for cron ${cronEntry.name} is $nextExecutionTime" }
+    cronQueue.add(CronQueueEntry(nextExecutionTime.toInstant(), cronEntry))
+  }
+
+  fun runReadyCrons() {
+    val now = clock.instant()
+    if (now.epochSecond - (lastCronRunAt?.epochSecond ?: 0L) > REBUILD_QUEUE_REQUIRED_SECS) {
+      logger.info { "Rebuilding cron queue required" }
+      rebuildQueue()
+    }
+
+    logger.info {
+      "Looking for ready crons - ${cronQueue.size} entries to scan, current time=$now"
+    }
+
+    while (cronQueue.isNotEmpty()) {
+      val cronQueueEntry = cronQueue.peek()
+      if (cronQueueEntry.executeAt > now) break
+
+      cronQueue.remove()
+      val cronEntry = cronQueueEntry.cronEntry
+      logger.info {
+        "CronJob ${cronEntry.name} was ready at " +
+          "${ZonedDateTime.ofInstant(cronQueueEntry.executeAt, zoneId)}"
+      }
+
+      enqueueCronJob(cronEntry)
```

Comment:
I see, so when the entry from the cronQueue is ready, we enqueue it in a job queue and process it on a handler thread. Nice.